File: repository/src/main/java/by/rekuts/travelagency/repository/impl/CountryRepositoryImpl.java
package by.rekuts.travelagency.repository.impl;

import by.rekuts.travelagency.domain.Country;
import by.rekuts.travelagency.repository.CountryRepository;
import by.rekuts.travelagency.repository.CountrySpecification;
import by.rekuts.travelagency.repository.Specification;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.criteria.Predicate;
import java.util.List;

@Repository
public class CountryRepositoryImpl implements CountryRepository {
	@PersistenceContext
	private EntityManager entityManager;

	@Override
	public void insert(Country country) {
		entityManager.persist(country);
	}

	@Override
	public void update(Country country) {
		var dbCountry = entityManager.find(Country.class, country.getId());
		dbCountry.setName(country.getName());
		entityManager.merge(dbCountry);
	}

	@Override
	public void delete(int id) {
	    entityManager.remove(entityManager.find(Country.class, id));
	}

    @Override
	public List<Country> getList(Specification specification) {
		final var countrySpecification = (CountrySpecification) specification;

		var builder = entityManager.getCriteriaBuilder();
		var criteriaQuery = builder.createQuery(Country.class);
		var countryRoot = criteriaQuery.from(Country.class);
        var predicates = countrySpecification.getPredicates(countryRoot, builder);

        if (!predicates.isEmpty()) {
            criteriaQuery.where(
                    predicates.toArray(new Predicate[]{})
            );
        }
        return entityManager.createQuery(criteriaQuery).getResultList();
	}
}

File: repository/src/main/scala/by/rekuts/travelagency/repository/impl/CountryRepositoryImpl.scala
package by.rekuts.travelagency.repository.impl

import java.util

import by.rekuts.travelagency.domain.Country
import by.rekuts.travelagency.repository.{CountryRepository, CountrySpecification, Specification}
import javax.persistence.criteria.Predicate
import org.springframework.stereotype.Repository

@Repository
class CountryRepositoryImpl extends CountryRepository {

  override def update(country: Country): Unit = {
    val dbCountry = entityManager.find(classOf[Country], country.id)
    dbCountry.name_$eq(country.name)
    entityManager.merge(dbCountry)
  }

  override def delete(id: Int): Unit = entityManager.remove(entityManager.find(classOf[Country], id))

  override def getList(specification: Specification[_]): util.List[Country] = {
    val countrySpecification = specification.asInstanceOf[CountrySpecification]
    val builder = entityManager.getCriteriaBuilder
    val criteriaQuery = builder.createQuery(classOf[Country])
    val countryRoot = criteriaQuery.from(classOf[Country])
    val predicates = countrySpecification.getPredicates(countryRoot, builder)
    if (!predicates.isEmpty) {
      criteriaQuery.where(predicates.toArray[Predicate](new Array[Predicate](predicates.size())): _*)
    }
    entityManager.createQuery(criteriaQuery).getResultList
  }
}
